axelrod/classifier.py

```diff
@@ -0,0 +1,236 @@
+        all_player_dicts[p.name] = new_player_dict
+
+    with open(filename, "w") as f:
+        yaml.dump(all_player_dicts, f)
+
+
+class _Classifiers(object):
+    """A singleton used to calculate any known classifier.
+
+    Attributes
+    ----------
+    all_player_dicts: A local copy of the dict saved in the classifier table.
+        The keys are player names, and the values are 'classifier' dicts (keyed
+        by classifier name).
+    """
+
+    _instance = None
+    all_player_dicts = dict()
+
+    # Make this a singleton
+    def __new__(cls):
+        if cls._instance is None:
+            cls._instance = super(_Classifiers, cls).__new__(cls)
+            # When this is first created, read from the classifier table file.
+            # Get absolute path
+            dirname = os.path.dirname(__file__)
+            filename = os.path.join(dirname, ALL_CLASSIFIERS_PATH)
+            with open(filename, "r") as f:
+                cls.all_player_dicts = yaml.load(f, Loader=yaml.FullLoader)
+
+        return cls._instance
+
+    @classmethod
+    def known_classifier(cls, classifier_name: Text) -> bool:
+        """Returns True if the passed classifier_name is known."""
+        global all_classifiers
+        return classifier_name in (c.name for c in all_classifiers)
+
+    @classmethod
+    def __getitem__(
+        cls, key: Union[Classifier, Text]
+    ) -> Callable[[Union[Player, Type[Player]]], Any]:
+        """Looks up the classifier for the player.
+
+        Given a passed classifier key, return a function that:
+
+        Takes a player.  If the classifier is found in the 'classifier' dict on
+        the player, then return that.  Otherwise look for the classifier for the
+        player in the all_player_dicts.  Returns None if the classifier is not
+        found in either of those.
+
+        The returned function expects Player instances, but if a Player class is
+        passed, then it will create an instance by calling an argument-less
+        initializer.  If no such initializer exists on the class, then an error
+        will result.
+
+        Parameters
+        ----------
+        key: A classifier or classifier name that we want to calculate for the
+            player.
+
+        Returns
+        -------
+        A function that will map Player (or Player instances) to their value for
+            this classification.
+        """
+        # Key may be the name or an instance.  Convert to name.
+        if not isinstance(key, str):
+            key = key.name
+
+        if not cls.known_classifier(key):
+            raise KeyError("Unknown classifier")
+
+        def classify_player_for_this_classifier(
+            player: Union[Player, Type[Player]]) -> Any:
+            # If the passed player is not an instance, then try to initialize an
+            # instance without arguments.
+            if not isinstance(player, Player):
+                try:
+                    player = player()
```

Comment:
Let's change the name of one of these to make it less confusing in the future. Maybe the function variable name can be `_player`, and then in the except clause here you could use `player = _player` if it can't be instanced without arguments (but I think that should usually work).

Reply:
I'm not sure what to do here.

My intention was to only allow classify_player_for_this_classifier to run on Player instances.  This is because for a strategy, a non-instance classifier dict will be stored in the yaml file, so it should be able to read this anyway.  (Actually I fixed the code just now to actually do this lookup for a non-trivial initializer.)  This code "if not isinstance..." is to simplify a lot of code that was doing trivial initializers, then passing in; it also allows filters and maps in some additional places.

I could do what you're asking for.  This would not change the existing behavior, and not cause a failure if the user passes a strategy without a trivial initializer.  But I almost want to throw a failure there, because it's better to be clear that this function is calling an instance.  We have for example code somewhere that says "[s for s in axelrod.strategies if Classifiers["stochastic"](s)]", but there are lots of strategies (like Random) which are stochastic **sometimes**.  I kind of don't want to give them a True or a False there; neither answer is right, and I want them to rewrite it to pass an instance.

On the other hand, it's weird that I'm essentially swallowing this error when I can trivially initialize.  And if somebody tried to make a strategy without a trivial initializer and add it to all_strategies, this would cause an unexpected error.  I'll probably end up doing what you're suggesting, but want to think about it a little more.

Reply:
Well maybe the solution is to require trivial initializers throughout the library, and/or throw a warning here that the default initializer was used, which a few strategies like memory-one currently do. Otherwise the user has to initialize their strategies before calling the `Classifier`, we can only read their minds so much.

If a change messes with one of the Meta strategies our tests should catch that. There's a test for example that tries to find violations when strategies claim to be not stochastic.

Reply:
I like that.  I implemented it that way.  It will raise an exception if there is not a trivial initializer.  Maybe we can enforce that _before_ Classifiers runs into it.